passenger_car_dynamic_model/src/passenger_car_dynamic_model/PassengerCarDynamicModel.cpp

```diff
@@ -0,0 +1,283 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include <stdlib.h>
+#include <math.h>
+#include <sstream>
+#include <functional>
+#include "passenger_car_dynamic_model/PassengerCarDynamicModel.h"
+
+/**
+ * Cpp containing the implementation of PassengerCarDynamicModel
+ */
+
+using namespace lib_vehicle_model;
+
+PassengerCarDynamicModel::PassengerCarDynamicModel() {
+  // Bind the callback functions
+  ode_func_ = std::bind(&PassengerCarDynamicModel::DynamicCarODE, this, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4);
+  post_step_func_ = std::bind(&PassengerCarDynamicModel::ODEPostStep, this, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4, std::placeholders::_5);
+};
+
+PassengerCarDynamicModel::~PassengerCarDynamicModel() {};
+
+void PassengerCarDynamicModel::setParameterServer(std::shared_ptr<ParameterServer> parameter_server) {
+  param_server_ = parameter_server;
+  
+  // Load Parameters
+  bool l_f_param        = param_server_->getParam("length_to_f", l_f_);
+  bool l_r_param        = param_server_->getParam("length_to_r", l_r_);
+  bool R_f_param        = param_server_->getParam("effective_wheel_radius_f", R_ef_);
+  bool R_r_param        = param_server_->getParam("effective_wheel_radius_r", R_er_);
+  bool long_stiff_param = param_server_->getParam("tire_longitudinal_stiffness", C_sx_);
+  bool lat_stiff_param  = param_server_->getParam("tire_cornering_stiffness", C_ay_);
+  bool inertia_param    = param_server_->getParam("moment_of_inertia", I_z_);
+  bool mass_param       = param_server_->getParam("vehicle_mass", m_);
+
+  // Check if all the required parameters could be loaded
+  if (!(l_f_param && l_r_param && R_f_param && R_r_param
+         && long_stiff_param && lat_stiff_param && inertia_param && mass_param)) {
+
+    std::ostringstream msg;
+    msg << "One of the required parameters could not be found or read " 
+      << " length_to_f: " << l_f_param 
+      << " length_to_r: " << l_r_param 
+      << " effective_wheel_radius_f: " << R_f_param 
+      << " effective_wheel_radius_r: " << R_r_param 
+      << " tire_longitudinal_stiffness: " << long_stiff_param
+      << " tire_cornering_stiffness: " << lat_stiff_param 
+      << " moment_of_inertia: " << inertia_param
+      << " vehicle_mass: " << mass_param;
+
+    throw std::invalid_argument(msg.str());
+  }
+}
+
+std::vector<VehicleState> PassengerCarDynamicModel::predict(const VehicleState& initial_state,
+  double timestep, double delta_t) {
+
+    // Populate control inputs
+    // This predict function takes in no new control inputs so extract the old ones from the state vector
+    VehicleControlInput control_input;
+    control_input.target_steering_angle = initial_state.prev_steering_cmd;
```

Comment:
I am still confusing about this. If we did not provide any control input, it probably means that we do not know the control inputs. Do you think it makes sense to use zero values as the control inputs?

Reply:
Since the previous command is in the state vector we can assume that CARMA will always provide it as needed. Therefore as long as we propagate it forward during prediction it will always contain valid values.